tests/test_logsumexp.py

```diff
@@ -5,24 +5,74 @@
 from timemachine.lib import custom_ops
 
 
+@pytest.mark.memcheck
+@pytest.mark.parametrize("precision", [np.float64, np.float32])
+def test_segmented_cuda_logsumexp_validation(precision):
+    max_values_per_segment = 5
+    num_segments = 2
+    if precision == np.float32:
+        summer = custom_ops.SegmentedLogSumExp_f32(max_values_per_segment, num_segments)
+    else:
+        summer = custom_ops.SegmentedLogSumExp_f64(max_values_per_segment, num_segments)
+
+    with pytest.raises(RuntimeError, match="number of segments don't match"):
+        summer.sum([[1.0]])
+
+    with pytest.raises(RuntimeError, match="empty array not allowed"):
+        summer.sum(
+            [
+                [
+                    1.0,
+                ],
+                [],
+            ]
+        )
+
+    with pytest.raises(RuntimeError, match="total values is greater than buffer size"):
+        summer.sum([[1.0] * (max_values_per_segment + 1) for _ in range(num_segments)])
+
+
 @pytest.mark.memcheck
 @pytest.mark.parametrize("precision,atol,rtol", [(np.float64, 1e-8, 1e-8), (np.float32, 2e-5, 1e-5)])
-@pytest.mark.parametrize("seed", list(range(2)))
+@pytest.mark.parametrize("seed", [2024])
 @pytest.mark.parametrize("loc", [1000, 0, -1000])
-@pytest.mark.parametrize("size", [2, 100, 10000])
-def test_cuda_logsumexp(precision, atol, rtol, seed, loc, size):
+@pytest.mark.parametrize("shape", [(1, 2), (100, 100), (1000, 1000)])
+def test_segmented_cuda_logsumexp(precision, atol, rtol, seed, loc, shape):
     rng = np.random.default_rng(seed)
 
-    values = rng.normal(size=size, loc=loc) * 1000.0
+    values = rng.normal(size=shape, loc=loc) * 1000.0
+
+    if precision == np.float32:
+        summer = custom_ops.SegmentedLogSumExp_f32(shape[1], shape[0])
+    else:
+        summer = custom_ops.SegmentedLogSumExp_f64(shape[1], shape[0])
+
+    test_vals = summer.sum(values)
+    assert len(test_vals) == shape[0]
+    np.testing.assert_array_equal(test_vals, summer.sum(values))
+
+    for test_val, vals in zip(test_vals, values):
+        np.testing.assert_allclose(logsumexp(vals), test_val)
+
 
-    reference_val = logsumexp(values)
+@pytest.mark.memcheck
+@pytest.mark.parametrize("precision,atol,rtol", [(np.float64, 1e-8, 1e-8), (np.float32, 2e-5, 1e-5)])
+@pytest.mark.parametrize("seed", [2024])
+@pytest.mark.parametrize("loc", [1000, 0, -1000])
+@pytest.mark.parametrize("num_samples", [2, 5, 10, 100])
+def test_segmented_cuda_logsumexp_ragged_arrays(precision, atol, rtol, seed, loc, num_samples):
+    rng = np.random.default_rng(seed)
+    max_values_per_segment = 100
+    values = [rng.normal(size=rng.integers(1, max_values_per_segment), loc=loc) * 1000.0 for _ in range(num_samples)]
```

Comment:
Currently 1) and 3) pass, 2) and 4) do not. We could do the same behavior that Scipy does to handle this (https://github.com/scipy/scipy/blob/main/scipy/special/_logsumexp.py#L107-L110) in the host (would make tests pass, but wouldn't change the exchange movers), where you might have +inf/-inf (jax potentials), or could do it in the GPU but that would be an additional kernel for code that would typically have nans because of how we handle overflows (and even if you converted to +inf/-inf, they are all indistinguishable in the device code)

> if one particular segment is bad

Do you mean does having a nan/inf in one segment break the other segments?

Reply:
> 2) and 4) do not.

What do they currently return?

> Do you mean does having a nan/inf in one segment break the other segments?

Yes, (assuming the other segments do not have a inf/nan)

Reply:
> What do they currently return?

They currently return Nan, which in the exchange move kernels we convert to +inf (https://github.com/proteneer/timemachine/blob/master/timemachine/cpp/src/kernels/k_exchange.cu#L49)

Reply:
Added test (does not match up with 2. and 4.) in https://github.com/proteneer/timemachine/pull/1239/commits/17124265ffa8903feff1e5392911f6c74e94e634